keps/sig-storage/2605-option-to-preserve-default-file-mode-bit-in-atomicwriter-volumes/README.md

```diff
@@ -0,0 +1,181 @@
+    // such as fsGroup etc
+    // +optional
+    DefaultMode *int32
+    // Specify whether the Secret or its key must be defined
+    // +optional
+    Optional *bool
+    // Specify whether we want to preserve the DefaultMode set on the files
+    // at the time of file creation.
+    // If PreserveDefaultMode is set to true the file mode is not modified based
+    // on fsGroup after the file is created with the defaultMode.
+    // If the PreserveDefaultMode is set to false and fsGroup is specified
+    // the volume is modified to be owned by the fsGroup.
+    // If PreserveDefaultMode is not specified in the volume spec, it defaults to false.
+    // +optional
+    PreserveDefaultMode *bool
+}
+```
+
+A secret volume source with preserveDefaultMode=true:
+
+```yaml
+  volumes:
+  - name: test-secret
+    secret:
+      preserveDefaultMode: true
+      defaultMode: 256
+      secretName: test-secret
+```
+
+### File Permission
+
+For the volumes that use AtomicWriter today the files are created with the Mode set in the KeyToPath struct.
+If the Mode is not specified then the DefaultMode set for the source is used.
+If DefaultMode is not specified then the file is created with mode bit set to 0600.
+If pod have fsGroup specified then all the files in the volume are walked,
+chown and chmod to the fsGroup
+
+#### Proposed heuristics
+
+- *Case 1*: The volume has PreserveDefaultMode set to false or not specified.
+    In this case there is no behavioural change in how the file permissions are
+    set for the volume. In other words, we will continue to create the files
+    with Mode/Default and then set the volume ownership based on fsGroup if specified.
+    This is the default case.
+- *Case 2*: The volume has PreserveDefaultMode set to true.
+    In this case the payload/volume/files are created using Mode/Default mode.
+    Then the call to SetVolumeOwnership is skipped, in turn preserving the Default file mode on the files.
+
+#### Alternatives considered
+
+- Instead of having PreserveDefaultMode flag at the volume source level, we can have the same flag in
+    the KeyToPath struct. That will allow us granular control over every file with-in the volume :
+  - We would have to pass the payload information along with PreserveDefaultMode to the SetVolumeOwnership routine
+        and skip chown/chmod if PreserveDefaultMode is set to true for the file.
+  - This will be more invasive change, but would allow more granular control at per file level
+  - If users want to preserve Mode for one file in the volume and not for another they can achieve that by creating
+        two different volumes, with PreserveDefaultMode set to true for the volume where they want to preserve default mode for all files
+
+### Scope Of the Change
+
+With this change we will try to update the behavior for below AtomicWriter volumes
+
+- SecretVolumeSource
+- DownwardAPIVolumeSource
+- ConfigMapVolumeSource
+- ProjectedVolumeSource
```

Comment:
At some point we will need to add CSI inline ephemeral volumes (`CSIVolumeSource`) to the mix, as they can provide ConfigMap/Secret-like volumes.

CSI driver vendors have option to set `CSIDriver.Spec.FSGroupPolicy = "None"` to opt out the whole CSI driver from fsGroup, but there is no per pod / per volume flag.

In addition, `CSIVolumeSource` does not have `DefaultMode`, we either need to add it there or the flag `PreserveDefaultMode` needs a different name.